pfcpiface/aether.go

```diff
@@ -0,0 +1,849 @@
+
+	log.Tracef("Created new module %v of type %v", moduleName, moduleClass)
+
+	return nil
+}
+
+func (a *aether) deleteModule(ctx context.Context, moduleName string) error {
+	req := &pb.DestroyModuleRequest{
+		Name: moduleName,
+	}
+
+	resp, err := a.client.DestroyModule(ctx, req)
+
+	if err != nil {
+		log.Errorf("DestroyModule RPC failed with err: %v\n", err)
+		return err
+	}
+
+	if resp.GetError() != nil && resp.GetError().Code != 0 {
+		log.Errorf("DestroyModule request '%+v' failed with err: %v\n", req, resp.GetError())
+		return status.Error(codes.Code(resp.GetError().Code), resp.GetError().Errmsg)
+	}
+
+	log.Tracef("Destroyed module %v", moduleName)
+
+	return nil
+}
+
+func (a *aether) sessionStats(pc *PfcpNodeCollector, ch chan<- prometheus.Metric) (err error) {
+	// TODO: implement
+	log.Traceln("sessionStats are not implemented in aether pipeline")
+	return
+}
+
+func (a *aether) setupBpfRules() (err error) {
+	ctx, cancel := context.WithTimeout(context.Background(), Timeout)
+	defer cancel()
+
+	// Do not modify. Hard-coded gates from aether.bess pipeline.
+	const (
+		ueTrafficPassGate     = 0
+		signalTrafficPassGate = 8192 - 1 // MAX_GATE - 1
+	)
+
+	// Pass-through filter for GTPU UE traffic.
+	ueFilter := "ip and dst host " + a.ownIp.IP.String() + " and udp dst port 2152"
+	if err = a.addBpfRule(ctx, ueFilter, -ueTrafficPassGate, ueTrafficPassGate); err != nil {
+		return
+	}
+
+	return
+}
+
+func (a *aether) pauseBessWorkers(ctx context.Context) error {
+	resp, err := a.client.PauseAll(ctx, &pb.EmptyRequest{})
+	if err != nil || resp.GetError() != nil {
+		log.Errorf("PauseAll rpc failed with resp: %v, err: %v\n", resp, err)
+		return err
+	}
+
+	return nil
+}
+
+func (a *aether) resumeBessWorkers(ctx context.Context) error {
+	resp, err := a.client.ResumeAll(ctx, &pb.EmptyRequest{})
+	if err != nil || resp.GetError() != nil {
+		log.Errorf("ResumeAll rpc failed with resp: %v, err: %v\n", resp, err)
+		return err
+	}
+
+	return nil
+}
+
+func (a *aether) addBpfRule(ctx context.Context, filter string, priority, gate int64) error {
+	f := pb.BPFArg_Filter{
+		Priority: priority,
+		Filter:   filter,
+		Gate:     gate,
+	}
+	bpfArg := &pb.BPFArg{Filters: []*pb.BPFArg_Filter{&f}}
+
+	err := a.processBpf(ctx, bpfArg, moduleMethodAdd)
+	if err != nil {
+		log.Errorln(err)
+		return err
+	}
+
+	return nil
+}
+
+func (a *aether) processBpf(ctx context.Context, msg proto.Message, method string) error {
+	switch method {
+	case moduleMethodAdd:
+		fallthrough
+	case moduleMethodDelete:
+		fallthrough
+	case moduleMethodClear:
+		fallthrough
+	case moduleMethodGetInitialArg:
+	default:
+		return ErrInvalidArgumentWithReason("method", method, "invalid method name")
+	}
+
+	any, err := anypb.New(msg)
+	if err != nil {
+		log.Error("Error marshalling the rule", msg, err)
+		return err
+	}
+
+	// BPF module is not thread-safe, need to pause processing.
+	if err = a.pauseBessWorkers(ctx); err != nil {
+		log.Errorln(err)
+		return err
+	}
```

Comment:
If `ModuleCommand` fails, BESS workers will be paused forever, so any further operations will be blocked until BESS is restarted (right?). I'm not sure if that's the case that we should be worried about, but the below suggestion (or its variation) shouldn't hurt:

```suggestion
	// BPF module is not thread-safe, need to pause processing.
	if err = a.pauseBessWorkers(ctx); err != nil {
		log.Errorln(err)
		return err
	}
	
	defer func() {
	    if err = a.resumeBessWorkers(ctx); err != nil {
		log.Fatal(err)
	    }
	}
```